### PR TITLE
DROOLS-3958 Unexpected Conflicting rows on rules analysis using date-effective and date-expires on Guided Decision Tables

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/DROOLS-3958

PRs:
https://github.com/kiegroup/drools/pull/2324
https://github.com/kiegroup/drools-wb/pull/1131

### CODE REVIEW

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java [hunk 1]
```diff
@@ -64,6 +78,20 @@ public CellUpdateManagerBase(final Index index,
...
+                    return true;
+                } else {
+                    isAttribute = true;
+                    break;
+                }
+            }
+        }
+        if (isAttribute) {
+            return false;
+        }
```
Comment (jomarko) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java]: I think could be simplified slightly, or am I missing something?
```java
public boolean update() {

        for (Map.Entry<String, RuleAttribute> entry : ruleAttributes.entrySet()) {
            if (coordinate.getCol() == entry.getValue().getIndex()) {
                if (updateRelevantAttribute(entry.getKey())) {
                    return true;
                } else {
                    return false;
                }
            }
        }
        
```

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                                           realCellValue);
+
+        final Comparable value = ActionBuilder.getValue(realCellValue);
+
+        if (attribute.equals(DateExpiresRuleAttribute.NAME)) {
+            final Date date = toDate(configuration,
+                                     value);
+            if (date != null) {
+                return Optional.of(new DateExpiresRuleAttribute(columnIndex,
+                                                                (Date) value));
```
Comment (jomarko) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java]: Shouldn't we use `new DateExpiresRuleAttribute(columnIndex, date)` ?

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            }
+        } else if (attribute.equals(DateEffectiveRuleAttribute.NAME)) {
+            final Date date = toDate(configuration,
+                                     value);
+            if (date != null) {
+                return Optional.of(new DateEffectiveRuleAttribute(columnIndex,
+                                                                  (Date) value));
```
Comment (jomarko) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java]: Shouldn't we use `new DateEffectiveRuleAttribute(columnIndex, date)` ?

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/DecisionTableAnalyzerUpdateTest.java [hunk 4]
```diff
@@ -86,6 +88,62 @@ public void testRowValueChange() throws Exception {
...
+                .buildTable();
+
+        fireUpAnalyzer();
+
+        assertContains(analyzerProvider.getAnalysisReport(),
+                       CheckType.REDUNDANT_ROWS,
+                       Severity.WARNING,
+                       1, 2);
+
+        setValue(1,
```
Comment (jomarko) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/DecisionTableAnalyzerUpdateTest.java]: Shouldn't we update the second row to have two valid intervals?

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/cache/util/DTableUpdateManagerAttributesTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        coordinates.add(coordinate);
+        table52.getData()
+                .get(0)
+                .get(2)
+                .setDateValue(null);
+
+        updateManager.update(table52,
+                             coordinates);
+
+        verify(analyzer).update(anySet());
```
Comment (jomarko) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/cache/util/DTableUpdateManagerAttributesTest.java]: I think this test is not very effective if we do not check the content of the `set` passed to the method `update`.

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/cache/util/DTableUpdateManagerAttributesTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        coordinates.add(coordinate);
+        table52.getData()
+                .get(0)
+                .get(3)
+                .setDateValue(null);
+
+        updateManager.update(table52,
+                             coordinates);
+
+        verify(analyzer).update(anySet());
```
Comment (jomarko) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/cache/util/DTableUpdateManagerAttributesTest.java]: I think this test is not very effective if we do not check the content of the `set` passed to the method `update`.

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/fromfile/EffectiveDatesFromFileTest.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import org.drools.verifier.api.reporting.Issue;
+import org.drools.verifier.core.main.Analyzer;
+import org.drools.workbench.models.guided.dtable.backend.GuidedDTXMLPersistence;
+import org.drools.workbench.models.guided.dtable.shared.model.GuidedDecisionTable52;
+import org.drools.workbench.services.verifier.plugin.client.AnalyzerUpdateTestBase;
+import org.drools.workbench.services.verifier.plugin.client.UpdateException;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static junit.framework.TestCase.assertTrue;
```
Comment (jomarko) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/fromfile/EffectiveDatesFromFileTest.java]: Can we please use `import static org.junit.Assert.assertTrue;` ?

**Comment by jomarko** [comment]
@Rikkola manual check passed, however would like to see responses to my comments before approving.

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java (hunk 1 above)
Comment (Rikkola) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java]: You might not be, but I remember I had to play with this a bit and I did try a simple solution, but this was the one that worked. So afraid to touch it ATM.

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java (hunk 2 above)
Comment (Rikkola) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java]: true

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java (hunk 3 above)
Comment (Rikkola) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/builders/AttributeBuilder.java]: true

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/DecisionTableAnalyzerUpdateTest.java (hunk 4 above)
Comment (Rikkola) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/DecisionTableAnalyzerUpdateTest.java]: The index numbers might be misleading, but this change results in both ranges being valid (date-effective is before date-expires) and the ranges do not overlap. So Date with 5000 becomes 1500.

**Comment by Rikkola** [comment]
I did the changes as a separate PR just to point out that only tests were mainly affected.

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java (hunk 1 above)
Comment (manstis) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java]: @jomarko Your proposal skips the remainder of the method. See [here](https://github.com/kiegroup/drools-wb/blob/e9b0fc416a8c0f0525115b19cc305d37940b7728/drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java#L96).

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/AnalyzerUpdateTestBase.java [hunk 8]
```diff
@@ -79,6 +80,23 @@ public ValueSetter setCoordinate() {
...
+        DTCellValue52 dtCellValue52 = table52.getData()
+                .get(rowIndex)
+                .get(columnIndex);
+        dtCellValue52
+                .setDateValue(value);
+        try {
+            updateManager.update(table52,
+                                 getUpdates(rowIndex,
+                                            columnIndex));
+        } catch (UpdateException e) {
```
Comment (manstis) [inline drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/AnalyzerUpdateTestBase.java]: This _smells_ bad.. what was your reasoning to catch `UpdateException`?

Comment (Rikkola) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/test/java/org/drools/workbench/services/verifier/plugin/client/AnalyzerUpdateTestBase.java]: I removed the UpdateException. I think it used to catch illegal states, but now it only checked if the cell can be found.

**Comment by Rikkola** [comment]
@manstis @jomarko the PR is updated and ready for reviews.

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java (hunk 1 above)
Comment (jomarko) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/CellUpdateManagerBase.java]: @manstis sorry, do not agree this time. In the removed code, the for cycle set `isAttribute` to `true` and break the cycle. Immediately after the cycle we have `if` that return `false` from the method if the `isAttribute` was true. However am not going to fight here. Approving.

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers Please merge as one commit.

**Pull request merged**